src/indexer.c

```diff
@@ -604,30 +604,8 @@ static void hash_partially(git_indexer *idx, const uint8_t *data, size_t size)
 
 static int write_at(git_indexer *idx, const void *data, off64_t offset, size_t size)
 {
-	git_file fd = idx->pack->mwf.fd;
-	size_t mmap_alignment;
-	size_t page_offset;
-	off64_t page_start;
-	unsigned char *map_data;
-	git_map map;
-	int error;
-
-	assert(data && size);
-
-	if ((error = git__mmap_alignment(&mmap_alignment)) < 0)
-		return error;
-
-	/* the offset needs to be at the mmap boundary for the platform */
-	page_offset = offset % mmap_alignment;
-	page_start = offset - page_offset;
-
-	if ((error = p_mmap(&map, page_offset + size, GIT_PROT_WRITE, GIT_MAP_SHARED, fd, page_start)) < 0)
-		return error;
-
-	map_data = (unsigned char *)map.data;
-	memcpy(map_data + page_offset, data, size);
-	p_munmap(&map);
-
+	if (p_pwrite(idx->pack->mwf.fd, data, size, offset) != (ssize_t)size)
+		return -1;
```

Comment:
I like that you've noticed that there's a posix call that matches our intent here - our posix indirection layer is really meant to use the actual posix functions when they're available, or implement a suitable fallback when they're not.  Can we do that here?

ie, can we just use `pwrite` when it's available?  I think that it's possible that there's no system that both has `pwrite` and lacks `mmap`, but the reason that I ask isn't because I think that the posix function is necessarily super useful for us.  What it _does_ do is act as a baseline in tests to ensure that _our_ `pwrite` implementation acts the same.  This is important because we expect `p_some_posix_fn` to act like `some_posix_fn`.  If it _doesn't_ then we're going to make errors when we call it because this is really meant to be a compatibility layer (at least as much as is possible).

(If our `p_pwrite` function really _isn't_ like `pwrite` then we should just keep calling it `write_at` or something.)

Anyway, this is clever, I wouldn't have thought of emulating `pwrite`.

Reply:
I have implemented p_pwrite using pwrite or p_mmap in src/posix.c since mmap based IO was used prior to this change. If there are systems where mmap based IO is more performant, we will not cause a regression.

I will replace the call of p_pwrite in write_at using a loop to avoid the typecasting of ssize_t to size_t.

Please let me know if I am missing something.